Add feature to define findings manually in the model, similar to over…

nozmore opened the pull request:
…rides

The idea here is after working with a dev team and threats are identified which does not exist in the pytm threatlib they can be manually added to the model. Similar to using overrides to modify a threat/finding. Alternatively someone could use pytm only for model definition, dfd/seq generation and reporting using manual findings instead of using any of the threatlib.

I originally called these "custom findings" and named variables accordingly. I was thinking of renaming these to "manual findings", thoughts? 

I would like to update tm.py and the README.md with a real world example that makes sense given the base tm model. I haven't spent any mental energy to come up with one that good and doesn't exist in the threatlib. If anyone has any thoughts let me know.

raphaelahrens commented:
Hi,

I think the notation is a little bit redundant
````
web.custom_findings = [
    Finding(web,
        id = "foo_id_3",
        description = "foo_description",
        details = "foo_details",
        severity = "HIGH",
        mitigations = "foo_mitigations",
        example = "foo_example",
        threat_id = "foo_threat_id_3",
        references = "foo_references",
        response = "accepted",
    ),
]
````
could just be something like
````
f(
  web,
  "foo_description",
  details = "foo_details",
  severity = "HIGH",
  mitigations = "foo_mitigations",
  example = "foo_example",
  threat_id = "foo_threat_id_3",
  references = "foo_references",
  response = "accepted",
)
````

So a function or class constructor could take care of adding the new Finding to the custom/manual_finding of `web`.
This removes the double use `web` and the list notation.
It can also create an id for us if we don't want to specify one.
It can also have required fields, like the description, directly as parameters.

If you prefer to have the element in the beginning of the line maybe 

````
web.add_finding( ...
````


Regarding the name I tend towards manual, but I'm not sure if there should be a separation between manual and automatic findinds.

nozmore commented:
Yes, I meant to mention that in the description, the redundancy is another reason I had this as draft. We haven't used methods for assignment in tm.py before so I just went with the same approach and obviously this doesn't work for json import. At minimum I was going to add logic to remove the dependency requirement. 

I think the required fields are up to the report format, not sure if we should impose this. For an internal threat maybe from a custom threat library it may just be a pointer or reference to a ticket.

I was thinking instead of treating them as separate findings in the report I could add them to the main findings list and the findings instance variables but add a 'source' field to the Finding (default would be 'pytm', source could be 'manual' for these findings or a custom value set in the model. To the earlier point when I copy them from the custom_findings instance variable to findings I could assign the target element to remove the redundancy.

raphaelahrens commented:
Source could also be an author, so that you have someone you can contact?

colesmj commented:
A couple of comments on this thread:

> It would potentially be useful to know the source of a finding as automatic (from pytm detection) or manual (human added) or possibly others like external or inherited.

> Findings should have common fields such as summary, description (long or short), etc, and should probably also contain references or some field to support a contact url or email or something (maybe a link to a support or security team, etc)

> To @nozmore 's comment about tying fields to the report, I'm a big fan of delinking. The findings should have metadata or data to be meaningful, the report can use what it wants/needs from findings, as long as there is a common set of fields and logic to detect when something is empty or missing.

nozmore commented:
> Source could also be an author, so that you have someone you can contact?

Exactly 'pytm' when found using the threatlib,  'manual' if the finding is defined in the model but 'source' is not set. any string (except for 'pytm') if 'source' is defined in the model.

colesmj commented:
> > Source could also be an author, so that you have someone you can contact?
> 
> > It would potentially be useful to know the source of a finding as automatic (from pytm detection) or manual (human added) or possibly others like external or inherited.
> 
> Exactly 'pytm' when found using the threatlib, 'manual' if the finding is defined in the model but 'source' is not set. any string (except for 'pytm') if 'source' is defined in the model.

Question: why should the finding(s) be defined _in_ the model, and not _with_ the model?

nozmore commented:
> Findings should have common fields such as summary, description (long or short), etc, and should probably also contain references or some field to support a contact url or email or something (maybe a link to a support or security team, etc)

> To @nozmore 's comment about tying fields to the report, I'm a big fan of delinking. The findings should have metadata or data to be meaningful, the report can use what it wants/needs from findings, as long as there is a common set of fields and logic to detect when something is empty or missing.

I re-used the Finding class so they do have a common set of fields however for these manually identified findings I don't believe requiring fields is for us to say. For threatlib findings, yes, fully agree. For these I think the organization maintaining their report template would need to decide what fields are required. Currently target is required and should be, adding description is reasonable so there is a way to describe the finding or point to something. 

I'm open to making whatever fields as required that we all agree to, I just don't think its necessary and will lead to people putting junk data like I did to satisfy the requirements of the tool.

colesmj commented:
> 
> I'm open to making whatever fields as required that we all agree to, I just don't think its necessary and will lead to people putting junk data like I did to satisfy the requirements of the tool.

I'm not suggesting making fields mandatory unless they make sense - summary, description, source (how finding got into the list) certainly, and an identifier (auto generated or manually supplied); all other fields should be optional, and if possible not fixed. The report should not rely on fields existing otherwise those fields should always exist and have at least some form of default (even if an empty string), or the report generator should know how to handle missing/empty fields in findings.

nozmore commented:
Ok. I think I incorporated all of the comments. Only thing I didn't was @colesmj comment about do the findings need to be stored in the model. No its code so they dont have to be, for now I like that it is part of the model. For me teams would manage 1 tm file per app (or per a given scope if a larger app) run this file thru the tool and produce output. Also there is a link to the model since the findings are added to elements. When I take a look at json support I would like to separate the input (model, manual findings) from the output (all findings) 

Below are the comment I addressed..

- Renamed to manual findings instead of custom
- manual findings are now incorporated with all other findings with a source field, (if separation is required the `if` template feature can be used)
- similar numerical finding 'id' counter is used
- the duplicate element reference in tm.py is removed `(web.manual_findings = [Finding(web........`
- removed required fields from Finding. 

The only outstanding change is to finalize on the custom findings in tm.py.... or decide to remove those from tm.py and just document them in the README.md.

nineinchnick commented on pytm/pytm.py:
```diff
@@ -1131,6 +1150,7 @@ class Element:
         doc="""Minimum TLS version required.""",
     )
     findings = varFindings([], doc="Threats that apply to this element")
+    manual_findings = varFindings([], doc="Findings manually added to this element")
```
Is it possible to populate the findings attribute manually, so adding a new attribute is not necessary?

nozmore replied:
Yes, thanks to the last round of changes I made to merge them in for reporting and have a 'source' attribute I can just use findings.

nineinchnick commented on pytm/pytm.py:
```diff
@@ -59,12 +59,18 @@ def __set__(self, instance, value):
-                "cannot overwrite {}.{} value with {}, already set to {}".format(
-                    instance, self.__class__.__name__, value, self.data[instance]
+            if (not isinstance(instance, Finding) 
+                or (isinstance(instance, Finding) and isinstance(self,varString) and self.data[instance] != "invalid")
```
I'm against using special values like "invalid". Looks like we should not use this `var*` type for `Findings`? What's exactly the purpose of these conditions?

nozmore replied:
invalid was already there, I didn't remove it. It set when a Finding is created without and element, I dont think it ever happened before because the element was required but actually gets used now. The conditions are used to workaround the 'overwrite' protection. I wanted to narrow down the specific occurances where overwrite would be allowed and is only for Finding's element and target attributes.

nineinchnick replied:
Maybe we can avoid setting the findings two times? Can you locate the place where it's being set the second time?

nozmore replied:
ok, cleaned things up.

nineinchnick commented on pytm/pytm.py:
```diff
@@ -626,6 +632,7 @@ class Finding:
 """,
     )
     cvss = varString("", required=False, doc="The CVSS score and/or vector")
+    source = varString("manual", required=False, doc="The source of the Finding.")
```
It's fine to have that logic, it's better to have this field only in `Finding` than both in here and in `Threat`

nineinchnick commented on tests/output.json:
```diff
@@ -801,4 +801,4 @@
     "threatsExcluded": [],
     "threatsFile": "pytm/threatlib/threats.json"
-}
+}
```
Please don't remove end of line at end of file. Configure your text editor not to do that automatically.
